src/lib/run-dev.js

```diff
@@ -0,0 +1,165 @@
+    } catch (e) {
+      aioLogger.error('unexpected error while cleaning up!')
+      aioLogger.error(e)
+      process.exit(1) // eslint-disable-line no-process-exit
+    }
+  })
+
+  try {
+    // TODO: Is there a chance of using run cmd on actions from plugin-runtime in future?
+    // In that case this backend stuff might have to go to lib-runtime ?
+
+    // Build Phase
+    if (withBackend) {
+      if (isLocal) {
+        const { config: localConfig, cleanup: localCleanup } = await runDevLocal(config, log, options.verbose)
+        devConfig = localConfig
+        cleanup.add(() => localCleanup(), 'cleaning up runDevLocal')
+      } else {
+        // check credentials
+        rtLibUtils.checkOpenWhiskCredentials(devConfig)
+        log('using remote actions')
+      }
+
+      // build and deploy actions
+      log('building actions..')
+      await buildActions(devConfig)
+
+      log(`watching action files at ${devConfig.actions.src}...`)
+      const { cleanup: watcherCleanup } = await actionsWatcher({ config: devConfig, isLocal, log })
+      cleanup.add(() => watcherCleanup(), 'stopping action watcher...')
+    }
+
+    if (hasFrontend) {
+      let urls = {}
+      if (config.app.hasBackend) {
+        // inject backend urls into ui
+        // note the condition: we still write backend urls EVEN if skipActions is set
+        // the urls will always point to remotely deployed actions if skipActions is set
+        log('injecting backend urls into frontend config')
+        urls = await rtLibUtils.getActionUrls(devConfig, true, isLocal && !skipActions)
+      }
+      await utils.writeConfig(devConfig.web.injectedConfig, urls)
+
+      if (!options.skipServe) {
+        const script = await utils.runPackageScript('build-static')
+        if (!script) {
+          const { cleanup: bundlerCleanup } = await bundle(devConfig, bundleOptions, log)
+          cleanup.add(() => bundlerCleanup(), 'cleaning up bundle...')
+        }
+      }
+    }
+
+    // Deploy Phase
+    if (withBackend || (hasFrontend && !options.skipServe)) {
+      // deploy actions
+      if (withBackend) {
+        log('redeploying actions..')
+        await deployActions(devConfig, isLocal, log)
+      }
+
+      // serve UI
+      if (hasFrontend) {
+        if (!options.skipServe) {
+          const script = await utils.runPackageScript('serve-static')
+          if (!script) {
+            const { url, cleanup: serverCleanup } = await serve(devConfig, bundleOptions, log)
+            frontEndUrl = url
+            cleanup.add(() => serverCleanup(), 'cleaning up serve...')
+          }
+        }
+      }
+    }
+
+    log('setting up vscode debug configuration files...')
+    const vscodeConfig = vscode(devConfig)
+    await vscodeConfig.update({ hasFrontend, withBackend, frontEndUrl })
+    cleanup.add(() => vscodeConfig.cleanup(), 'cleaning up vscode debug configuration files...')
+
+    if (!isLocal && !hasFrontend) {
```

Comment:
Isn't this only required if `( !isLocal && !hasFrontend ) && ( !config.app.hadBackend || !fetchLogs )` 
Should we move this condition to after the pollerCleanup setup?

Reply:
Seems confusing to add that logic. I'll add a boolean (`needsProcessWaiter`) to set to false (default true): if it's not set to false, we run the dummy proc.

It is only set to false when:
1.  run local (OW running) OR
2. front-end http server is running OR
3. log poller is running (probably skip this set for now, since it's always true from the app run command, without a flag to not fetch logs)

Reply:
fixed in f39c724